src/main/java/org/candlepin/subscriptions/marketplace/MarketplaceApiFactory.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (c) 2019 Red Hat, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ * Red Hat trademarks are not licensed under GPLv3. No permission is
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+package org.candlepin.subscriptions.marketplace;
+
+import org.candlepin.subscriptions.http.HttpClient;
+import org.candlepin.subscriptions.marketplace.api.resources.MarketplaceApi;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.FactoryBean;
+import org.springframework.stereotype.Component;
+
+/**
+ * Factory that produces marketplace API clients.
```

Comment:
Is this a pattern that's being used elsewhere in the application?

Reply:
Yeah, many places, I tried to refactor as much common code out as possible. See `org.candlepin.subscriptions.http.HttpClient`. Unfortunately, the generated clients don't use base classes for ApiClient or for MarketplaceApi (or similar), which makes it difficult to DRY any further. If you have ideas, happy to try them though.